core/src/streams.rs

```diff
@@ -378,114 +383,147 @@ impl<'gc> NetStream<'gc> {
+                    stream_format: swf_format,
+                    num_samples_per_block: 0,
+                    latency_seek: 0,
+                };
 
-                    match audio_data.data {
-                        FlvAudioDataType::Raw(data)
-                        | FlvAudioDataType::AacSequenceHeader(data)
-                        | FlvAudioDataType::AacRaw(data) => {
-                            substream.append(slice.to_subslice(data))?
-                        }
-                    };
+                *audio_stream = Some((substream, sound_stream_head));
 
-                    if context.is_action_script_3() {
-                        Ok(context
-                            .audio
-                            .start_substream(substream.clone(), &sound_stream_head)?)
-                    } else if let Some(mc) = attached_to {
-                        Ok(context.audio_manager.start_substream(
-                            context.audio,
-                            substream.clone(),
-                            mc,
-                            &sound_stream_head,
-                        )?)
-                    } else {
-                        return Err(NetstreamError::NotAttached);
-                    }
-                })();
+                &mut audio_stream.as_mut().unwrap().0
+            }
+        };
 
-                if let Err(e) = audio_handle {
-                    tracing::error!("Error encountered starting stream: {}", e);
+        Ok(substream.append(data)?)
+    }
+
+    /// Determine if the given sound is currently playing.
+    fn sound_currently_playing(
+        context: &mut UpdateContext<'_, 'gc>,
+        sound: &Option<SoundInstanceHandle>,
+    ) -> bool {
+        sound
+            .map(|si| context.audio.is_sound_playing(si))
+            .unwrap_or(false)
+    }
+
+    /// Clean up after a sound instance that has finished playing.
+    ///
+    /// Generally speaking, streams are only to be used once. However, the
+    /// audio backend will only retain information about sounds that are
+    /// currently playing, so if the sound has finished since the last tick, we
+    /// need to restart it.
+    ///
+    /// Intended to be called at the start of tag processing, before any new
+    /// audio data has been streamed.
+    fn collect_sound_stream(
+        self,
+        context: &mut UpdateContext<'_, 'gc>,
+        write: &mut NetStreamData<'gc>,
+    ) {
+        if !Self::sound_currently_playing(context, &write.sound_instance) {
+            write.audio_stream = None;
+            write.sound_instance = None;
+        }
+    }
+
+    /// Ensure that if we have queued up audio into a sound stream, that said
+    /// stream gets sent over to the audio backend.
+    ///
+    /// Intended to be called at the end of tag processing. Audio processing
+    /// should occur only after a minimum number of tags have been processed to
+    /// avoid audio underruns.
```

Comment:
What determines this "minimum number" of tags to process? Is it the same as `max_lookahead_audio_tags`?

Reply:
There's no particular criteria, it just has to be 'enough'. And in practice it will be `max_lookahead_audio_tags` unless there's something strange going on. When I implement progressive download we may need to enforce this.